### PR TITLE
Master mail owl canned responses jgi

### PR DESCRIPTION
**Pull request by jgi-odoo**

### CODE REVIEW

File: addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="mail.CannedResponseSuggestion" owl="1">
+        <a class="o_CannedResponseSuggestion dropdown-item" t-att-class="{ 'active': props.isActive }" href="#" role="menuitem" t-att-data-id="cannedResponse.id"
+           t-on-click="_onClick" t-on-mouseover="_onMouseOver">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml]: no new line between attributes in this case

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 2]
```diff
@@ -177,6 +177,213 @@ QUnit.test('add emoji replaces (keyboard) text selection', async function (asser
...
+        '.o_ComposerTextInput_mentionDropdownPropositionList',
+        "canned responses suggestions list should not be present"
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`).focus();
+    document.execCommand('insertText', false, ":");
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keydown'));
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keyup'));
+    await afterNextRender();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer_tests.js]: wrap code trigger render inside

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 3]
```diff
@@ -177,6 +177,213 @@ QUnit.test('add emoji replaces (keyboard) text selection', async function (asser
...
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`).focus();
+    await afterNextRender(() =>
+        document.execCommand('insertText', false, ":")
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keydown'));
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keyup'));
+    await afterNextRender();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer_tests.js]: same

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 4]
```diff
@@ -177,6 +177,213 @@ QUnit.test('add emoji replaces (keyboard) text selection', async function (asser
...
+        "text content of composer should have content"
+    );
+    await afterNextRender(() =>
+        document.execCommand('insertText', false, ":")
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keydown'));
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keyup'));
+    await afterNextRender();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer_tests.js]: same

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 5]
```diff
@@ -177,6 +177,213 @@ QUnit.test('add emoji replaces (keyboard) text selection', async function (asser
...
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`).focus();
+    await afterNextRender(() =>
+        document.execCommand('insertText', false, ":")
+    );
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keydown'));
+    document.querySelector(`.o_ComposerTextInput_textarea`)
+        .dispatchEvent(new window.KeyboardEvent('keyup'));
+    await afterNextRender();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer_tests.js]: same

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 6]
```diff
@@ -177,6 +177,213 @@ QUnit.test('add emoji replaces (keyboard) text selection', async function (asser
...
+        "Hello Odoo ",
+        "text content of composer should have previous content + canned response substitution + additional whitespace afterwards"
+    );
+
+    // select emoji
+    await afterNextRender(() =>
+        document.querySelector('.o_Composer_buttonEmojis').click()
+    );
+    document.querySelector('.o_EmojisPopover_emoji[data-unicode="😊"]').click();
+    await afterNextRender();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer_tests.js]: same

**Review by seb-odoo — COMMENTED** [review]
Looks extremely similar to mentions, so looks ok in general.

I'm wondering if there isn't "too much" copy/paste, especially in regard to the display of the list and the handle of keyboard navigation.

Looks like we might want a generic component working on a generic list of models instead. All the computation based on index are exactly the same and the model itself is irrelevant to them.

File: addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml (hunk 1 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml]: Done

File: addons/mail/static/src/components/composer/composer_tests.js (hunk 2 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/composer/composer_tests.js]: Done

File: addons/mail/static/src/components/composer/composer_tests.js (hunk 3 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/composer/composer_tests.js]: Done

File: addons/mail/static/src/components/composer/composer_tests.js (hunk 4 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/composer/composer_tests.js]: Done

File: addons/mail/static/src/components/composer/composer_tests.js (hunk 5 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/composer/composer_tests.js]: Done

File: addons/mail/static/src/components/composer/composer_tests.js (hunk 6 above)
Comment (jgi-odoo) [reply addons/mail/static/src/components/composer/composer_tests.js]: Done

**Comment by jgi-odoo** [comment]
Based on comments, I also changed tests related to mention manager

File: addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml]: change isn't pushed?

Comment (jgi-odoo) [reply addons/mail/static/src/components/canned_response_suggestion/canned_response_suggestion.xml]: It is

**Comment by seb-odoo** [comment]
Need to delete this PR, rebase on 13.4 and create new PR on /odoo repo (+ update PR link in task)

**Pull request closed without merging**